src/test/socket/bind/test_bind.rs

```diff
@@ -0,0 +1,588 @@
+fn test_double_bind_loopback_and_any(
+    reverse: bool,
+    sock_type: libc::c_int,
+    flag: libc::c_int,
+) -> Result<(), String> {
+    let fd1 = unsafe { libc::socket(libc::AF_INET, sock_type | flag, 0) };
+    assert!(fd1 >= 0);
+    let fd2 = unsafe { libc::socket(libc::AF_INET, sock_type | flag, 0) };
+    assert!(fd2 >= 0);
+
+    let addr1 = libc::sockaddr_in {
+        sin_family: libc::AF_INET as u16,
+        sin_port: 11111u16.to_be(),
+        sin_addr: libc::in_addr {
+            s_addr: libc::INADDR_LOOPBACK.to_be(),
+        },
+        sin_zero: [0; 8],
+    };
+
+    let addr2 = libc::sockaddr_in {
+        sin_family: libc::AF_INET as u16,
+        sin_port: 11111u16.to_be(),
+        sin_addr: libc::in_addr {
+            s_addr: libc::INADDR_ANY.to_be(),
+        },
+        sin_zero: [0; 8],
+    };
+
+    // if reverse, bind to ANY before LOOPBACK
+    let (addr1, addr2) = if reverse {
+        (addr2, addr1)
+    } else {
+        (addr1, addr2)
+    };
+
+    let args1 = BindArguments {
+        fd: fd1,
+        addr: Some(LibcSockAddr::In(addr1)),
+        addr_len: None,
+    };
+
+    let args2 = BindArguments {
+        fd: fd2,
+        addr: Some(LibcSockAddr::In(addr2)),
+        addr_len: None,
+    };
+
+    run_and_close_fds(&[fd1, fd2], || {
+        check_bind_call(&args1, false, None)?;
+        check_bind_call(&args2, true, Some(libc::EADDRINUSE))?;
+        Ok(())
+    })?;
+
+    Ok(())
+}
+
+// test binding to all interfaces with a port of 0
+fn test_unspecified_port(sock_type: libc::c_int, flag: libc::c_int) -> Result<(), String> {
+    let fd = unsafe { libc::socket(libc::AF_INET, sock_type | flag, 0) };
+    assert!(fd >= 0);
+
+    let addr = libc::sockaddr_in {
+        sin_family: libc::AF_INET as u16,
+        sin_port: 0u16.to_be(),
+        sin_addr: libc::in_addr {
+            s_addr: libc::INADDR_ANY.to_be(),
+        },
+        sin_zero: [0; 8],
+    };
+
+    let args = BindArguments {
+        fd: fd,
+        addr: Some(LibcSockAddr::In(addr)),
+        addr_len: None,
+    };
+
+    run_and_close_fds(&[fd], || check_bind_call(&args, false, None))?;
+
+    Ok(())
+}
+
+// run the function and then close any given file descriptors, even if there was an error
+fn run_and_close_fds<F>(fds: &[libc::c_int], f: F) -> Result<(), String>
+where
+    F: Fn() -> Result<(), String>,
+{
+    let rv = f();
+
+    for fd in fds.iter() {
+        let fd = *fd;
+        let rv_close = unsafe { libc::close(fd) };
+        assert_eq!(rv_close, 0, "Could not close the fd");
+    }
+
+    rv
+}
+
+fn get_errno() -> i32 {
+    std::io::Error::last_os_error().raw_os_error().unwrap()
+}
+
+fn get_errno_message(errno: i32) -> String {
+    let cstr;
+    unsafe {
+        let error_ptr = libc::strerror(errno);
+        cstr = std::ffi::CStr::from_ptr(error_ptr)
+    }
+    cstr.to_string_lossy().into_owned()
+}
+
+fn check_bind_call(
+    args: &BindArguments,
+    should_error: bool,
```

Comment:
I originally made it this way since you might want to say that something fails, but you don't care about the errno. I did this with the expectation that different versions of glibc might have changed the errno at some point (since some of the errnos, especially when there are multiple different errors, are arbitrary), but in practice this didn't seem to be the case for the socket() or bind() calls. I'll remove this from these tests.